### ISSUE TITLE
Empty output when executing query with ALLOW FILTERING

### ISSUE THREAD

**Issue by JerryKwan** [scylladb/scylladb#11340]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [x] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): 5.0.1-0.20220719.b177dacd3
Cluster size: 3
OS (RHEL/CentOS/Ubuntu/AWS AMI): CentOS

when executing CQL with ALLOW FILTERING n cqlsh, it only output "---MORE---" but without any records details. but if execute the same query with all of the partition keys specified without "ALLOW FILTERING", it will show the records

**Comment by tzach** [comment]
Can you please provide a full CQL statement?

**Comment by JerryKwan** [comment]
@tzach 
You can use the following CQL to do some tests
```
select * from records where source = 'a_sale' and date = '2022-08-22' and vin = '1FA6P8CF9N51230XX' ALLOW FILTERING;
```
and the schema of records is
```
CREATE TABLE records (
    source text,
    date date,
    update_time bigint,
    record_id text,
    details_hash text,
    type text,
    vin text,
    PRIMARY KEY ((source, date), update_time, record_id)
) WITH CLUSTERING ORDER BY (update_time ASC, record_id ASC);
```
The output of executing in cqlsh is "---MORE---", but when executing the same CQL using the cassandra's Python SDK, the output is something like {'source': 'a_sale', 'date': Date(19226), 'update_time': 1661155064163858, 'record_id': '5e6a56b3', 'details_hash': '7b2ee4b8f739c55035f02e72d562007a60e9564fbe02d54d2517050d656b7719', 'type': 'sale', 'vin': '1FA6P8CF9N51230XX'}, that's why I think there might be something wrong with cqlsh

**Comment by tzach** [comment]
I think this is relevant
https://docs.scylladb.com/stable/kb/cqlsh-more.html

**Comment by nyh** [comment]
@JerryKwan if you press enter on this "MORE" a few times, do you get results? If you never get any results in any page, this is an outright bug. But I suspect that eventually you will get a page with some results:

As the document that @tzach pointed to explains, if you have a lot of data in your database but the SELECT filter is very specific, Scylla may need to read a lot of data just to discover that none of it matches the filter and it shouldn't return any of it, and in that case it returns an empty page. If you continue the query (with the "MORE"), it will read some more data - and eventually find some row that you wanted to return.

If Cassandra does **not** do that, I think it is Cassandra's bug: Consider that you have a database of a billion rows that takes 10 minutes to read entirely, and you run a filter that is only supposed to match one single item. Will the query just hang for 10 minutes until returning one row? This won't work - the query will time out much earlier.

That being said, maybe the current behavior can be considered a bug not in Scylla but in cqlsh, or perhaps more accurately an inconvenience rather than a bug: Cqlsh could see that it got an empty page and automatically ask for the next page, without asking the user to type a newline.

**Comment by JerryKwan** [comment]
@tzach @nyh Thanks for looking into this.
Yes, after pressing several 'ENTER' keys, it show the desired record.
@nyh I agree with your comment "Cqlsh could see that it got an empty page and automatically ask for the next page, without asking the user to type a newline", cqlsh has a parameter named 'request-timeout', so I think it should automatically ask the next page, and stop asking more if timeout

**Comment by nyh** [comment]
@JerryKwan I guess it depends who "high level" cqlsh is supposed to be. Right now, it's pretty low-level - when paging is enabled, cqlsh simply sends the query to Scylla (or Cassandra) with paging enable - telling Scylla to stop after some (configurable) number of rows. Every time Scylla returns a page, cqlsh prints "More" and stop. There is currently a direct correspondence between pages returned by Scylla and pages shown by cqlsh.

Arguably, cqlsh's paging can be higher level and resemble the Unix "more" utility - if asked to show 100 results it can continue to fetch pages until 100 are available to be shown and then show exactly 100 (keeping any overflow results for the next page). Perhaps cqlsh can even use the Linux "less" instead of trying to implement paging, like (for example) "git log" does: cqlsh could continuously read more and more pages from Scylla outputting all the results into a pipe to "less". When the pipe buffer is full, cqlsh is paused during the write to the pipe, so will stop making additional page-read requests until the user continues to see more lines of output and the pipe becomes available for write again.

We could make less-based paging a new paging mode - e.g., "paging pager" (instead of "paging on"), using your default PAGER (e.g., "less").

The downside of my proposal is that it distances the cqlsh user from what actually happens in the lowlevel protocol. But what is happening in the low-level protocol (i.e., empty pages) is exactly what bothered you when you created this issue.

Related stackoverflow discussions:
https://stackoverflow.com/questions/34306773/is-there-a-way-to-display-results-in-a-pager-in-cassandra-cqlsh

**Comment by nyh** [comment]
It's worth mentioning that if this issue has become or will become a request to enhance cqlsh, then the fix will need to be in the repository holding cqlsh - not Scylla. I'm not sure if we have a Scylla for of cqlsh (@fee-mendes, do we?) or we will need to propose a fix to the upstream cqlsh, but in both cases this issue will need to move to a different repository.

**Comment by fee-mendes** [comment]
> I'm not sure if we have a Scylla for of cqlsh (@fee-mendes, do we?)

I don't think so. In particular, we currently can't use the latest CQLsh due to #11106 - so it is going to be hard to propose a fix to upstream.

**Comment by avelanarius** [comment]
> I'm not sure if we have a Scylla fork of cqlsh

I believe all of cqlsh code is in scylla-tools-java:

https://github.com/scylladb/scylla-tools-java/blob/master/bin/cqlsh.py
https://github.com/scylladb/scylla-tools-java/tree/master/pylib/cqlshlib

It  includes some Scylla-specific fixes, such as https://github.com/scylladb/scylla-tools-java/commit/f408ce81ba0da8594b72d71d3cf3591e23096d8e.

**Comment by JerryKwan** [comment]
@nyh Thanks for the explanation
If I remembered correctly, the output of cqlsh against Cassandra is different from cqlsh against ScyllaDB, when executing the same query against Cassandra, it output the result directly, there is no blank "--more--" output.

**Comment by nyh** [comment]
I opened an issue #12102 which is arguably a duplicate of this one, but explains the problem in a more Scylla-centric and less cqlsh-centric way, and explains how Scylla differs from Cassandra on the meaning of `fetch_size` (@JerryKwan also noted this above), and why it's not clear which definition is better.

But the two issues are not exact duplicates: If we decide in #12102 to change Scylla to interpret `fetch_size` the same way that Cassandra does, then this issue will also be closed as solved as well. But if in #12102 we decide to close that issue as WONTFIX - without changing Scylla - then we should still solve **this** issue by improving cqlsh to show constant-visual-size pages, regardless of the "pages" that Scylla returns, as explained in comments above.
